### ISSUE TITLE
NULLIF returns bool instead of correct type

### ISSUE THREAD

**Issue by kyleconroy** [sqlc-dev/sqlc#133]
### Version

1.14.0

### What happened?

Actually, NULLIF also smells bad. When using it in a query, the Go struct is generated with a field with a bool type, instead of a NullInt64 type.

### Database schema

```sql
CREATE TABLE author (
  id bigserial NOT NULL
);
```

### SQL queries

```sql
-- name: GetRestrictedId :one
SELECT
  NULLIF(id, $1) restricted_id
FROM
  author;
```

### Configuration

```yaml
{
  "version": "1",
  "packages": [
    {
      "engine": "postgresql",
      "path": "go",
      "name": "querytest",
      "schema": "query.sql",
      "queries": "query.sql"
    }
  ]
}
```

### Playground URL

https://play.sqlc.dev/p/321eb83b945edfab5a41d8c91feb596bcab9860ce1780639d1f9863207a20644

### What operating system are you using?

macOS

### What database engines are you using?

PostgreSQL

### What type of code are you generating?

Go

_Originally posted by @euller88 in https://github.com/kyleconroy/sqlc/issues/92#issuecomment-536583273_

**Comment by skabbes** [comment]
I think this is an endless game of whack-a-mole.  I _think_ in general type-checking is impossible, take for example:

```sql
select
  sqlc.arg('string_or_float')::int as x
```
how can sqlc know what type the first arg is?  It can be anything castable to int - including a string or floating point number, numeric(10, 2), etc.  I (personally) believe you need to give the developer control over this.


I think my proposal:
https://github.com/kyleconroy/sqlc/issues/1525 would at least give the developer control of this behavior, which would allow sqlc to still be usable (but less magic), and also would leave room for "adding magic" in the future. (because theoretically this individual case should be type-checkable).

```sql
-- name: GetRestrictedId :one
SELECT
  sqlc.output(
    NULLIF(id, sqlc.arg('ignore_id', 'type', 'bigint', 'nullable', false)),
   'nullable', true) as restricted_id
FROM
  author;
```

**Comment by kyleconroy** [comment]
This is fixed in [v1.23.0](https://docs.sqlc.dev/en/latest/reference/changelog.html#v1-23-0) by enabling the [database-backed query analyzer](https://docs.sqlc.dev/en/latest/howto/generate.html#enhanced-analysis-with-managed-databases). We added a [test case](https://github.com/sqlc-dev/sqlc/tree/main/internal/endtoend/testdata/null_if_type/postgresql/pganalyzer) for this issue so it won’t break in the future.

You can play around with the [working example on the playground](https://play.sqlc.dev/p/4454736f19c0b5f1a9f1a671291ddb287b6265cf2aef9a35d16491340b522092)

**Comment by Jonathan-Landeed** [comment]
I still get this issue in views.
